### PR TITLE
Add onCommitted listeners for transactions

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This should simply some of the idempotent development by allowing any PersistenceTransaction consumer to register onCommitted listeners.

### CODE REVIEW

File: packages/firestore/src/local/persistence.ts [hunk 1]
```diff
@@ -29,14 +29,25 @@ import { RemoteDocumentCache } from './remote_document_cache';
...
  * When you call Persistence.runTransaction(), it will create a transaction and
  * pass it to your callback. You then pass it to any method that operates
  * on persistence.
  */
 export abstract class PersistenceTransaction {
+  private readonly onCommittedListeners: Array<() => {}> = [];
+
   abstract readonly currentSequenceNumber: ListenSequenceNumber;
+
+  addOnCommittedListener(listener: () => {}): void {
```
Comment (wilhuff) [inline packages/firestore/src/local/persistence.ts]: Shouldn't we require that the committed listeners return `void`? That way you can't accidentally have a return statement in your callback that you intend for the outer method.

Or would that interfere with using short arrow functions in the callback?

File: packages/firestore/src/local/memory_persistence.ts [hunk 2]
```diff
@@ -180,6 +180,7 @@ export class MemoryPersistence implements Persistence {
     this.referenceDelegate.onTransactionStarted();
     return transactionOperation(txn)
       .next(result => {
+        txn.raiseOnCommittedEvent();
```
Comment (wilhuff) [inline packages/firestore/src/local/memory_persistence.ts]: This ordering seems off. Couldn't the reference delegate conceivably modify the transaction?

IndexedDbPersistence has the better implementation. Chaining this off the promise makes it clearer that it's altogether outside the transaction.

File: packages/firestore/src/local/persistence.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/persistence.ts]: I just realized this when I started to modify my "applyRemoteEvent" PR. It is much better if we return void.

File: packages/firestore/src/local/memory_persistence.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/memory_persistence.ts]: Changed this to invoke the committed handler as the last step.

**Comment by schmidt-sebastian** [comment]
Tests added. Whee.

File: packages/firestore/test/unit/local/persistence_transaction.test.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        const targetsStore = IndexedDbPersistence.getStore<
+          DbTargetKey,
+          { targetId: TargetId }
+        >(txn, DbTarget.store);
+
+        txn.addOnCommittedListener(() => {
+          ++commitCount;
+        });
+
+        ++runCount;
```
Comment (wilhuff) [inline packages/firestore/test/unit/local/persistence_transaction.test.ts]: Optional: you might also consider expecting that commitCount is zero during the transaction run.

Comment (schmidt-sebastian) [reply packages/firestore/test/unit/local/persistence_transaction.test.ts]: Done

**Pull request merged**